### PR TITLE
Implement Kve Op and Kv helper

### PR DESCRIPTION
**Pull request by ricardoV94**
Needed for https://github.com/pymc-devs/pymc-experimental/pull/389

Related to https://github.com/pymc-devs/pytensor/discussions/1038

CC @AuguB 

----
📚 Documentation preview 📚: https://pytensor--1081.org.readthedocs.build/en/1081/

### CODE REVIEW

File: pytensor/scalar/math.py [hunk 1]
```diff
@@ -1281,6 +1281,36 @@ def c_code(self, *args, **kwargs):
...
+        return scipy.special.kv(v, x)
+
+    def impl(self, v, x):
+        return self.st_impl(v, x)
+
+    def L_op(self, inputs, outputs, output_grads):
+        v, x = inputs
+        [out] = outputs
+        [g_out] = output_grads
+        # -(v / x) * kv(v, x) - kv(v - 1, x)
```
Comment (ricardoV94) [inline pytensor/scalar/math.py:1300]: There is a `scipy.kvp` for computing the derivative, but that would require another Op without counterparts in different backends

**Comment by dehorsley** [comment]
Is it worth exposing Kve as well? I think you can use it to prevent some underflow for large argument in your log in https://github.com/pymc-devs/pymc-experimental/pull/389

**Comment by ricardoV94** [comment]
> Is it worth exposing Kve as well? I think you can use it to prevent some underflow for large argument in your log in https://github.com/pymc-devs/pymc-experimental/pull/389

You think so, do we even need kv then or should it just be a helper around kve? Reducing core Ops is always nice

**Comment by dehorsley** [comment]
> You think so? 

`exp(x)` rounds to 0 around -700, any values less than that will give you nans in your logp.

> Do we even need kv then or should it just be a helper around kve? Reducing core Ops is always nice

Maybe not! Include a stabilising rewrite for Kv -> Kve * exp(-|x|)

**Comment by dehorsley** [comment]
I notice TFP also has [log_bessel_kve](https://www.tensorflow.org/probability/api_docs/python/tfp/math/log_bessel_kve) which should be more numerically stable again in this use case. That's probably a good candidate as a rewrite target.

**Comment by ricardoV94** [comment]
> Maybe not! Include a stabilising rewrite for Kv -> Kve * exp(-|x|)

Kv = Kve * exp(x) not -x. I'm not sure it's like the Iv/Ive combo.

There must be cases where kv is preferred otherwise why would it be offered? If it's never preferred there's no reason to have it offered.

Re log, I've also came across a paper that implements a stable form iteratively (see linked discussion), but seems like quite some complexity. I would wait until there's demand for it.

**Comment by ricardoV94** [comment]
The scipy page shows an example of underflowing earlier, so you're probably right. https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.kve.html

Is kv more stable for smaller x, or does kve win across the whole range. If so I'll just ditch kv altogether

**Comment by dehorsley** [comment]
> > Maybe not! Include a stabilising rewrite for Kv -> Kve * exp(-|x|)
> 
> Kv = Kve * exp(x) not -x. I'm not sure it's like the Iv/Ive combo.

Other way around is it not? Kv ~ exp(-x)/sqrt(z) for large z, so Kve = Kv * exp(x) is the "non exponential" part 

> 
> There must be cases where kv is preferred otherwise why would it be offered? If it's never preferred there's no reason to have it offered.

I'm sure there must be but at least since we're mostly dealing with logps, I'd say we'll want to factor out the exponential term most of the time for stability and kve will probably get much more use. 

Maybe a good argument for having Kve and not Kv is someone without experience with numerical analysis is more likely to write something stable if they only have Kve 😀 

> 
> Re log, I've also came across a paper that implements a stable form iteratively (see linked discussion), but seems like quite some complexity. I would wait until there's demand for it.


For a log(kve) specialisation, we could use the TFP implementation (which claims to be more stable) in jax and fallback to the naive implementation in scipy/c. That said, for the range of values I tried with the TFP log(kve), I didn't see any difference between it and the naive implementation, so maybe not worth it.

**Comment by ricardoV94** [comment]
> Other way around is it not? Kv ~ exp(-x)/sqrt(z) for large z, so Kve = Kv * exp(x) is the "non exponential" part

I didn't think it through with 1/exp(x) = exp(-x) :P 

> For a log(kve) specialisation, we could use the TFP implementation (which claims to be more stable) in jax and fallback to the naive implementation in scipy/c. That said, for the range of values I tried with the TFP log(kve), I didn't see any difference between it and the naive implementation, so maybe not worth it.

Those are different things. JAX/TFP cannot be used instead of scipy, they are different backends.

But yes we can over kve, and kv as `kve * pt.exp(-x)`. Gonna try to run a small benchmark to see if it's indeed more stable across the mark

**Comment by ricardoV94** [comment]
Btw @dehorsley if you want to hang out in the devs discourse server here is a link: https://discord.gg/vYkmxNuF

Say hi when / if you join so we can grant you access to the private channels

**Comment by dehorsley** [comment]
> @dehorsley I switched so we have a core kve, and a helper kv based on kve. I also added a stabilization rewrite for `log(kv)`. What do you think?

Sounds good to me!

> 
> A specialization for `log(kve)` can be left for later if there's a demand for more stability. Good to know that there are options.

Sure. For reference, it looks like the TFP implementation actually calculates log(Kve)/log(Ive), then applies `exp` to get the Kve/Ive implementations. I'm not sure how good jax/XLA's optimisation pass is, but maybe it is able to simplify the `log(exp)` expression too. That might explain why I was seeing bit-for-bit identical outputs between `log(tfp.math.bessel_kve(nu, x))` and `tfp.math.log_bessel_kve(nu, x)`. 

Incidentally, it seems like TFP's Kve implementation is more numerically stable than SciPy's: `scipy.special.kve(1,x)` gives NaNs around x=1e10, whereas `tfp.math.bessel_kve(1,x)` is happy for x all the way up to the float limit. 

> 
> Should we do something similar for Iv/Ive (and remove the core Iv Op)? If so, I'll open an issue for this, it need not block this PR

Yeah I think so.

**Review by dehorsley — APPROVED** [review]
Looks good to me, just a couple of nit-picks 😁

**Comment by ricardoV94** [comment]
Thanks @dehorsley, nits addressed

**Pull request merged**